symengine/polynomial.h

```diff
@@ -8,9 +8,253 @@
 #include <symengine/expression.h>
 #include <symengine/monomials.h>
 
+// Calculates bit length of number, used in UIntDict*= only
+template <typename T>
+unsigned int bit_length(T t)
+{
+    unsigned int count = 0;
+    while (t > 0) {
+        count++;
+        t = t >> 1;
+    }
+    return count;
+}
+
 namespace SymEngine
 {
 
+class UIntDict
+{
+public:
+    //! Holds the dictionary for a UnivariateIntPolynomial
+    map_uint_mpz dict_;
+
+public:
+    UIntDict()
+    {
+    }
+
+    ~UIntDict() SYMENGINE_NOEXCEPT
+    {
+    }
+
+    UIntDict(const UIntDict &) = default;
+
+    UIntDict(UIntDict &&other) SYMENGINE_NOEXCEPT
+        : dict_(std::move(other.dict_))
+    {
+    }
+
+    UIntDict(const int &i)
+    {
+        if (i != 0)
+            dict_ = {{0, integer_class(i)}};
+    }
+
+    UIntDict(const map_uint_mpz &p)
+    {
+        dict_ = p;
+        auto iter = dict_.begin();
+        while (iter != dict_.end()) {
+            if (iter->second == integer_class(0)) {
+                auto toErase = iter;
+                iter++;
+                dict_.erase(toErase);
+            } else
+                iter++;
+        }
+    }
+
+    UIntDict(const integer_class &expr)
+    {
+        if (expr != integer_class(0))
+            dict_ = {{0, std::move(expr)}};
+    }
+
+    UIntDict &operator=(const UIntDict &) = default;
+
+    UIntDict &operator=(UIntDict &&other) SYMENGINE_NOEXCEPT
+    {
+        if (this != &other)
+            this->dict_ = std::move(other.dict_);
+        return *this;
+    }
+
+    friend UIntDict operator+(const UIntDict &a, const UIntDict &b)
+    {
+        UIntDict c = a;
+        c += b;
+        return c;
+    }
+
+    UIntDict &operator+=(const UIntDict &other)
+    {
+        for (auto &it : other.dict_) {
+            auto t = dict_.lower_bound(it.first);
+            if (t != dict_.end() and t->first == it.first) {
+                t->second += it.second;
+                if (t->second == 0) {
+                    dict_.erase(t);
+                }
+            } else {
+                dict_.insert(t, {it.first, it.second});
+            }
+        }
+        return *this;
```

Comment:
This method is very inefficient for an unordered map. You need to rewrite this whole method.